libmamba/include/mamba/specs/version.hpp

```diff
@@ -0,0 +1,202 @@
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_SPECS_VERSION_HPP
+#define MAMBA_SPECS_VERSION_HPP
+
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace mamba::specs
+{
+
+    /**
+     * A succession of a number and a  lowercase litteral.
+     *
+     * Comparison is done lexicographically, with the number first and the litteral second.
+     * Certain litterals have special meaning:
+     * "*" < "dev" < "_"< any other litteral < "" < "post"
+     */
+    class VersionPartAtom
+    {
+    public:
+
+        VersionPartAtom() noexcept = default;
+        VersionPartAtom(std::size_t numeral) noexcept;
+        VersionPartAtom(std::size_t numeral, std::string_view litteral);
+        // The use of a template is only meant to prevent ambiguous conversions
+        template <typename Char>
+        VersionPartAtom(std::size_t numeral, std::basic_string<Char>&& litteral);
+
+        auto numeral() const noexcept -> std::size_t;
+        auto litteral() const& noexcept -> const std::string&;
+        auto litteral() && noexcept -> std::string;
+
+        auto str() const -> std::string;
+
+        auto operator==(const VersionPartAtom& other) const -> bool;
+        auto operator!=(const VersionPartAtom& other) const -> bool;
+        auto operator<(const VersionPartAtom& other) const -> bool;
+        auto operator<=(const VersionPartAtom& other) const -> bool;
+        auto operator>(const VersionPartAtom& other) const -> bool;
+        auto operator>=(const VersionPartAtom& other) const -> bool;
+
+    private:
+
+        // Stored in decreasing size order for performance
+        std::string m_litteral = "";
+        std::size_t m_numeral = 0;
+    };
+
+    extern template VersionPartAtom::VersionPartAtom(std::size_t, std::string&&);
+
+    /**
+     * A sequence of VersionPartAtom meant to represent a part of a version (e.g. major, minor).
+     *
+     * Version parts can have a arbitrary number of atoms, such as {0, "post"} {1, "dev"}
+     * in 0post1dev
+     *
+     * @see  Version::parse for how this is computed from strings.
+     * @todo Use a small vector of expected size 1 if performance ar not good enough.
+     */
+    using VersionPart = std::vector<VersionPartAtom>;
+
+    /**
+     * A sequence of VersionPart meant to represent all parts of a version.
+     *
+     * CommonVersion are composed of an aribtrary postive number parts, such as major, minor.
+     * They are typically separated by dots, for instance the three parts in 3.0post1dev.4 are
+     * {{3, ""}}, {{0, "post"}, {1, "dev"}}, and {{4, ""}}.
+     *
+     * @see  Version::parse for how this is computed from strings.
+     * @todo Use a small vector of expected size 4 if performance ar not good enough.
+     */
+    using CommonVersion = std::vector<VersionPart>;
```

Comment:
So ok we need arbitrary number parts (assuming that's a requirement here) but having vectors of vectors feels like it could be simplified/flatten in some ways.
Would it be possible to use one vector of VersionPartAtom with a way to know the begin and end of each version part?

Reply:
That would however complicate the tests...

Reply:
Would complicate much the very custom comparisons as well. It would help to have some `std::span` and `std::ranges` for sure!

Or would an allocator aware design make this better?